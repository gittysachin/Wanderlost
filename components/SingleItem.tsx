import { gql } from '@apollo/client'
import styled from 'styled-components'
import Head from 'next/head'

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${(props) => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`

const SINGLE_ITEM_QUERY = gql`
  query GetItem($id: ID!) {
    item(id: $id) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`

function SingleItem({ item }) {
  return (
    <SingleItemStyles>
      <Head>
        <title>Wanderlost | {item.title}</title>
      </Head>
      <img src={item.largeImage} alt={item.title} />
      <div className="details">
        <h2>Viewing {item.title}</h2>
        <p>{item.description}</p>
      </div>
    </SingleItemStyles>
  )
}

export default SingleItem
export { SINGLE_ITEM_QUERY }
