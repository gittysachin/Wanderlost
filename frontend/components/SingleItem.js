import { useQuery, gql } from '@apollo/client';
import styled from 'styled-components';
import Head from 'next/head';
import Error from './ErrorMessage';

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${(props) => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      largeImage
    }
  }
`;

function SingleItem({ id }) {
  const { data, loading, error } = useQuery(SINGLE_ITEM_QUERY, {
    variables: { id },
  });
  if (loading) return <p>Loading...</p>;
  if (error) return <Error error={error} />;
  if (!data.item) return <p>No Item Found for {id}</p>;
  const { item } = data;
  return (
    <SingleItemStyles>
      <Head>
        <title>Sick Fits | {item.title}</title>
      </Head>
      <img src={data.item.largeImage} alt={item.title} />
      <div className="details">
        <h2>Viewing {item.title}</h2>
        <p>{item.description}</p>
      </div>
    </SingleItemStyles>
  );
}

export default SingleItem;
export { SINGLE_ITEM_QUERY };
