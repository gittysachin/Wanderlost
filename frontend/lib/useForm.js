import { useState } from 'react';

const uploadFile = async ({ value, inputs, updateInputs }) => {
  console.log('Uploading file');
  const data = new FormData();
  data.append('file', value);
  data.append('upload_preset', 'sickfits');

  const res = await fetch(
    'https://api.cloudinary.com/v1_1/davyhausser/image/upload',
    {
      method: 'POST',
      body: data,
    },
  );
  const file = await res.json();
  console.log(file);
  updateInputs({
    ...inputs,
    image: file.secure_url,
    largeImage: file.eager[0].secure_url,
  })
};

export default function useForm(initial = {}) {
  const [inputs, updateInputs] = useState(initial);

  function handleChange(e) {
    let { value, name, type } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      [value] = e.target.files;
      uploadFile({ value, inputs, updateInputs });
    }
    updateInputs({
      ...inputs,
      [name]: e.target.value,
    });
  }

  function resetForm() {
    updateInputs(initial);
  }

  return {
    inputs,
    handleChange,
    resetForm
  };
}