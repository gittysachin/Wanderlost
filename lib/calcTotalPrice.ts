import { GetCurrentUser_user_cart } from '../components/__generated__/GetCurrentUser'

export default function calcTotalPrice(
  cart: Array<GetCurrentUser_user_cart>
): number {
  return cart.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally
    return tally + cartItem.quantity * cartItem.item.price
  }, 0)
}
