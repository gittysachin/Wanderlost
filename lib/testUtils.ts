import '@testing-library/jest-dom/extend-expect'
import casual from 'casual'
import { GetItem_item } from '../components/__generated__/GetItem'
import { GetOrder_order, GetOrder_order_items } from '../components/__generated__/GetOrder'
import {
  GetCurrentUser_user,
  GetCurrentUser_user_cart,
} from '../components/__generated__/GetCurrentUser'

// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): GetItem_item => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
})

const fakeUser = (): Omit<GetCurrentUser_user, 'permissions'> => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  // permissions: [Permission.ADMIN],
  // orders: [],
  cart: [],
})

const fakeOrderItem = (): GetOrder_order_items => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = (): GetOrder_order => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
})

const fakeCartItem = (overrides: Partial<GetCurrentUser_user_cart>): GetCurrentUser_user_cart => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  // user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  private store: {
    [key: string]: string
  }

  constructor() {
    this.store = {}
  }

  clear(): void {
    this.store = {}
  }

  getItem(key: string): string | null {
    return this.store[key] || null
  }

  setItem(key: string, value: number): void {
    this.store[key] = value.toString()
  }

  removeItem(key: string): void {
    delete this.store[key]
  }
}

export { LocalStorageMock, fakeItem, fakeUser, fakeCartItem, fakeOrder, fakeOrderItem }
