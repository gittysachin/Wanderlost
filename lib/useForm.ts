import { useState, ChangeEvent, Dispatch, SetStateAction } from 'react'

type Inputs = Record<string, string>
type Event = ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
type Form = {
  inputs: Inputs
  setInputs: Dispatch<SetStateAction<Inputs>>
  handleChange: (event: Event) => void
  resetForm: () => void
  clearForm: () => void
}

async function uploadFile({ value, inputs, setInputs }) {
  const data = new FormData()
  data.append('file', value)
  data.append('upload_preset', 'sickfits')

  const res = await fetch('https://api.cloudinary.com/v1_1/davyhausser/image/upload', {
    method: 'POST',
    body: data,
  })
  const file = await res.json()
  setInputs({
    ...inputs,
    image: file.secure_url,
    largeImage: file.eager[0].secure_url,
  })
}

export default function useForm(initial = {}): Form {
  const [inputs, setInputs] = useState(initial)

  function handleChange(e: { target: { files?: any; name?: any; type?: any; value?: any } }) {
    const { name, type } = e.target
    let { value } = e.target
    if (type === 'number') {
      value = parseInt(value, 10)
    }
    if (type === 'file') {
      [value] = e.target.files
      uploadFile({ value, inputs, setInputs })
    }
    setInputs({
      ...inputs,
      [name]: value,
    })
  }

  function resetForm() {
    setInputs(initial)
  }

  function clearForm() {
    const blankState = Object.fromEntries(Object.entries(inputs).map(([key]) => [key, '']))
    setInputs(blankState)
  }

  return {
    inputs,
    setInputs,
    handleChange,
    resetForm,
    clearForm,
  }
}
