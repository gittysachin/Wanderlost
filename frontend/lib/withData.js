import withApollo from 'next-with-apollo';
import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
      request: (operation) => {
        operation.setContext({
          fetchOptions: {
            credentials: 'include',
          },
          headers,
        });
      },
    }),
    resolvers: {
      Mutation: {
        toggleCart(_, variables, { cache }) {
          // read the cartOpen value from the cache
          const { cartOpen } = cache.readQuery({
            query: LOCAL_STATE_QUERY,
          });
          // Write the cart state to the opposite
          const data = {
            data: { cartOpen: !cartOpen },
          };
          cache.writeData(data);
          return data;
        },
      },
    },
  });
}

export default withApollo(createClient);
