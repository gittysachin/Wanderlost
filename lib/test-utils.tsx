import React from 'react'
import { render as defaultRender, RenderResult } from '@testing-library/react'
import { RouterContext } from 'next/dist/next-server/lib/router-context'
import { NextRouter } from 'next/router'
import '@testing-library/jest-dom/extend-expect'
import casual from 'casual'

export * from '@testing-library/react'

import { GetItem_item } from '../components/__generated__/GetItem'
import {
  GetOrder_order,
  GetOrder_order_items,
} from '../components/__generated__/GetOrder'
import {
  GetCurrentUser_user,
  GetCurrentUser_user_cart,
} from '../components/__generated__/GetCurrentUser'

// --------------------------------------------------
// Override the default test render with our own
//
// You can override the router mock like this:
//
// const { baseElement } = render(<MyComponent />, {
//   router: { pathname: '/my-custom-pathname' },
// });
// --------------------------------------------------
type DefaultParams = Parameters<typeof defaultRender>
type RenderUI = DefaultParams[0]
type RenderOptions = DefaultParams[1] & { router?: Partial<NextRouter> }

export function render(
  ui: RenderUI,
  { wrapper, router, ...options }: RenderOptions = {}
): RenderResult {
  if (!wrapper) {
    wrapper = ({ children }) => {
      const Wrapper = (
        <RouterContext.Provider value={{ ...mockRouter, ...router }}>
          {children}
        </RouterContext.Provider>
      )
      return Wrapper
    }
  }

  return defaultRender(ui, { wrapper, ...options })
}

const mockRouter: NextRouter = {
  basePath: '',
  pathname: '/',
  route: '/',
  asPath: '/',
  query: {},
  push: jest.fn(),
  replace: jest.fn(),
  reload: jest.fn(),
  back: jest.fn(),
  prefetch: jest.fn(),
  beforePopState: jest.fn(),
  events: {
    on: jest.fn(),
    off: jest.fn(),
    emit: jest.fn(),
  },
  isFallback: false,
}

// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): GetItem_item => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
})

const fakeUser = (): Omit<GetCurrentUser_user, 'permissions'> => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  // permissions: [Permission.ADMIN],
  // orders: [],
  cart: [],
})

const fakeOrderItem = (): GetOrder_order_items => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = (): GetOrder_order => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
})

const fakeCartItem = (
  overrides: Partial<GetCurrentUser_user_cart>
): GetCurrentUser_user_cart => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  // user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  private store: {
    [key: string]: string
  }

  constructor() {
    this.store = {}
  }

  clear(): void {
    this.store = {}
  }

  getItem(key: string): string | null {
    return this.store[key] || null
  }

  setItem(key: string, value: number): void {
    this.store[key] = value.toString()
  }

  removeItem(key: string): void {
    delete this.store[key]
  }
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
}
