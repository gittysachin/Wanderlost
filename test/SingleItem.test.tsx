import { MockedProvider } from '@apollo/react-testing'
import { render, fakeItem } from '../lib/test-utils'
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem'

const item = fakeItem()
describe('<SingleItem/>', () => {
  test('renders with proper data', async () => {
    const mocks = [
      {
        // when someone makes a request with this query and variable combo
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        // return this fake data (mocked data)
        result: {
          data: {
            Item: item,
          },
        },
      },
    ]
    const { container, findByTestId } = render(
      <MockedProvider mocks={mocks}>
        <SingleItem item={item} />
      </MockedProvider>
    )
    await findByTestId('singleItem')
    expect(container).toMatchSnapshot()
  })

  // test('Errors with a not found item', async () => {
  //   const mocks = [
  //     {
  //       request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
  //       result: {
  //         errors: [{ message: 'Items Not Found!' }],
  //       },
  //     },
  //   ]
  //   const { container } = render(
  //     <MockedProvider mocks={mocks}>
  //     <SingleItem item={item} />
  //     </MockedProvider>
  //   )

  //   await screen.findByTestId('graphql-error')
  //   expect(container).toHaveTextContent('Items Not Found!')
  // })
})
