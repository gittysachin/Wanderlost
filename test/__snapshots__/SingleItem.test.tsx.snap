// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem/> renders with proper data 1`] = `
<div>
  <div
    class="sc-bdnylx lmZiGp"
    data-testid="singleItem"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        Viewing 
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
